frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -0,0 +1,1495 @@
+	pub fn is_signed(&self) -> bool {
+		matches!(self, Phase::Signed)
+	}
+
+	/// Weather the phase is unsigned or not.
+	pub fn is_unsigned(&self) -> bool {
+		matches!(self, Phase::Unsigned(_))
+	}
+
+	/// Weather the phase is unsigned and open or not, with specific start.
+	pub fn is_unsigned_open_at(&self, at: Bn) -> bool {
+		matches!(self, Phase::Unsigned((true, real)) if *real == at)
+	}
+
+	/// Weather the phase is unsigned and open or not.
+	pub fn is_unsigned_open(&self) -> bool {
+		matches!(self, Phase::Unsigned((true, _)))
+	}
+
+	/// Weather the phase is off or not.
+	pub fn is_off(&self) -> bool {
+		matches!(self, Phase::Off)
+	}
+}
+
+/// A configuration for the module to indicate what should happen in the case of a fallback i.e.
+/// reaching a call to `elect` with no good solution.
+#[cfg_attr(test, derive(Clone))]
+pub enum FallbackStrategy {
+	/// Run a on-chain sequential phragmen.
+	///
+	/// This might burn the chain for a few minutes due to a stall, but is generally a safe
+	/// approach to maintain a sensible validator set.
+	OnChain,
+	/// Nothing. Return an error.
+	Nothing,
+}
+
+/// The type of `Computation` that provided this election data.
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, RuntimeDebug)]
+pub enum ElectionCompute {
+	/// Election was computed on-chain.
+	OnChain,
+	/// Election was computed with a signed submission.
+	Signed,
+	/// Election was computed with an unsigned submission.
+	Unsigned,
+}
+
+impl Default for ElectionCompute {
+	fn default() -> Self {
+		ElectionCompute::OnChain
+	}
+}
+
+/// A raw, unchecked solution.
+///
+/// This is what will get submitted to the chain.
+///
+/// Such a solution should never become effective in anyway before being checked by the
+/// [`Pallet::feasibility_check`]
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug)]
+pub struct RawSolution<C> {
+	/// Compact election edges.
+	compact: C,
+	/// The _claimed_ score of the solution.
+	score: ElectionScore,
+	/// The round at which this solution should be submitted.
+	round: u32,
+}
+
+impl<C: Default> Default for RawSolution<C> {
+	fn default() -> Self {
+		// Round 0 is always invalid, only set this to 1.
+		Self { round: 1, compact: Default::default(), score: Default::default() }
+	}
+}
+
+/// A checked solution, ready to be enacted.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct ReadySolution<A> {
+	/// The final supports of the solution.
+	///
+	/// This is target-major vector, storing each winners, total backing, and each individual
+	/// backer.
+	supports: Supports<A>,
+	/// The score of the solution.
+	///
+	/// This is needed to potentially challenge the solution.
+	score: ElectionScore,
+	/// How this election was computed.
+	compute: ElectionCompute,
+}
+
+/// Solution size of the election.
+///
+/// This is needed for proper weight calculation.
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, RuntimeDebug, Default)]
+pub struct SolutionSize {
```

Comment:
I am kinda mixed about removing this and just re-using `RoundSnapshotMetadata`: They only differ in the `compact`-ness. One is for the data that we store in storage, one is something that the solution transaction submits (and is not stored in state in itself). 

Reused or not?

Reply:
Yeah, it really makes sense to me to do this. I am just not sure about the `#[codec(compact)]`. This should work fine with storage right? @thiolliere

Reply:
all in all I think we can use compact or not compact it shouldn't matter much as it is only few bytes. 
In case you use compact then it will use compact encoding in storage too, which is fine.

Reply:
addressed.